### ISSUE TITLE
Feature request: stop condition

### ISSUE THREAD

**Issue by fuzzah** [AFLplusplus/AFLplusplus#197]
Thank you for your awesome work guys!

I use AFL++ a lot and i think it would be great if it had optional condition to stop fuzzing jobs after certain amount of minutes (hours) without new paths discovered.

For example: stop all fuzzing jobs if **each** status screen shows at least 30 minutes without new paths.

**Comment by hexcoder-** [comment]
Thanks.
This could be done in afl-whatsup for all fuzzers running on the local machine, 
but in general not for distributed fuzzing.

Then afl-whatsup would get the additional feature to kill/stop all fuzzers.

**Comment by vanhauser-thc** [comment]
@hexcoder- I like that solution. we already have so many features in afl-fuzz that everything that we dont have to put in there is good :)

**Comment by domenukk** [comment]
Why not also for distributed fuzzing? Place some kill file once the condition triggers, every fuzzer can then stop individually at the next sync.

**Comment by hexcoder-** [comment]
@domenukk : I assume you mean
 1. place a kill file with terminating condition in the output dir before fuzzing starts
 2. let each fuzzer watch for the terminating condition and exit, when done

Sounds like a good idea, but needs more conceptual work (e.g. what conditions should be possible).
Also we already have limits for max. fuzzing time and max. number of executions as command line options. Any new functionality should be consistent to that IMHO.

**Comment by vanhauser-thc** [comment]
@hexcoder- no he means if one fuzzer is at the kill condition (e.g. 60 minutes without a new path), then the fuzzer places a file it its sync dir, e.g. sync_dir/Slave_5/KILL and other fuzzers check for that and then kill themselves when they see that.

that is IMHO however not a good thing. Only the fuzzer itself should kill, and not kill the whole fuzzing farm. simply because you run the instances with different strategies, and you would only self-kill instances that do not perform.

I thought about this and rather than a fuzzer killing itself it should rather inform like we do with the -I option on new crashes. @fuzzah what do you think?

**Comment by fuzzah** [comment]
@vanhauser-thc , option like -I could be fine too, but it's separate for each afl-fuzz instance :(
And it would be great if you could include sample script to stop fuzzing jobs in examples/ directory.
 
The whole reason behind this feature request is to stop fuzzing gracefully when 'enough' paths checked, but there's no definition of 'enough'. One way would be to calculate some sort of 'path discovery speed' and stop fuzzing if average path discovery speed is let's say less than 0.01 in last few hours.
 
Imagine this: you run fuzzing job for let's say a week. From time to time you check if it discovers new paths. Each time you check it you see something like 'last new path 1 hour ago', but you could miss that moment when it wasn't discovering new paths for 12 hours straight. We don't have any way to react to this.
Another situation: 4 afl-fuzz instances ran in parallel, 3 of them were not finding new paths for 24 hours, but the last one kept discovering new paths. Proposed option similar to -I/-E/-V could maybe stop those 3 "bad" instances, but better way would be for them to keep trying and wait for the last instance to stop finding new paths and only then they all would stop.
 
**Simply put**: we need a way to check if cycle counter in each status screen changed their color to blue, we need to capture this state and it would indicate that fuzzing job is done. How to capture this state? Maybe execute some user-provided script, or maybe just stop all fuzzing jobs with success message. Of course it's not about the color, we should let AFL++ know how much time to wait before letting us know fuzzing job is complete.

**Comment by domenukk** [comment]
My idea was actually as @hexcoder- interpreted it: afl-whatsup (or something else, I'd probably prefer this to be in the Master node) watches and places a kill file once the conditions triggers.
No new path in the last x hours would be one possible & sensible condition.

**Comment by vanhauser-thc** [comment]
OK I gave this some thought.

I understand the need to stop the fuzzer when it seems to make no sense to run longer. or at least detect that it is not useful to let it run further.

however when would that be? that is kinda up to the decision of the person running the fuzzing campaign, so this is not something we should code hard.

instead: we expose the necessary information. the missing information is `cycles_wo_finds` - the number of cycles that have not found any paths, and this is now in the fuzzer_stats output file. Now you can see when the last path was found, the cycles_wo_find etc for each fuzzer instance and then write a script to reach on this.

I will now add the entry to afl-whatsup and then I think this can be closed?

**Comment by vanhauser-thc** [comment]
@fuzzah I added it. I think now a script running afl-whatsup can decide if it should terminate the fuzzing campaign or not.

**Comment by vanhauser-thc** [comment]
@fuzzah please close the issue if you think its fine now

**Comment by fuzzah** [comment]
@vanhauser-thc , thank you!

Please take a look at this screenshot of my fuzzing situation:
![image](https://user-images.githubusercontent.com/60884276/77320397-8fac3f00-6d21-11ea-928f-49d245802077.png)
 
Still with knowledge of `cycles_wo_finds` I couldn't catch that moment when all fuzzing jobs have e.g. 3 hours of not finding paths.

**Comment by vanhauser-thc** [comment]
you do. per fuzzer afl-whatsup displays also when the last path was found.
so what information is still missing?

**Comment by fuzzah** [comment]
Created PR #271 
I think in future afl-whatsup could also have some sort of verbose mode or maybe output stats in CSV format, in which all the time intervals would be represented as unix timestamps or "number of seconds since event X" so that it would be easier to parse by third-party tools.
